Display shipping calculator link for guests shopper

tarunvijwani opened the pull request:
<!-- Please do not remove any information from this pull request. Instead, add N/A or leave blank if not applicable -->

## What

Fixes woocommerce/woocommerce-blocks#11435

## Why

- The shipping calculator link was displayed based on the `isAddressComplete` value.
- The shipping calculator was not displayed for the scenarios where the address was incomplete, but the formatted address was present.
- This made the shipping calculator not getting displayed for guest shoppers.
- With this, conditions are changed from `isAddressCompletea` to `formatShippingAddress` to display the shipping calculator if the formatted address is present.
- Added the unit test case for the condition for formatted address.

<!-- Describe the reason for your changes. This will help the reviewer and future readers get additional context -->

## Testing Instructions

### Case 1
1. Go to the WooCommerce settings page: `wp-admin/admin.php?page=wc-settings`.
2. Change 'Default customer location' to `No location by default`. 
3. Remove any default(fallback) shipping rates. 
4. Open your site in a private session(incognito mode).
5. Add some products to the cart. 
6. Go to the cart page, and confirm the "Add an address for shipping rates" link is present in the shipping calculator.
7. Click on the link, and confirm it is working as expected.
8. Open your site in a regular session where you're already logged in and have placed an order before. 
9. Add some products to the cart. 
10. Go to the cart page, confirm the shipping address, and "change address" link present in the shipping calculator.
11. Click on the link, and confirm it is working as expected.

### Case 2
1. Go to the WooCommerce settings page: `wp-admin/admin.php?page=wc-settings`.
2. Change 'Default customer location' to `No location by default`. 
3. Add any default(fallback) shipping rates. 
4. Open your site in a private session(incognito mode).
5. Add some products to the cart. 
6. Go to the cart page, and confirm the shipping calculator is hidden and default shipping rates are visible
7. Open your site in a regular session where you're already logged in and have placed an order before. 
8. Add some products to the cart. 
9. Go to the cart page, confirm the shipping address, and "change address" link present in the shipping calculator.
10. Click on the link, and confirm it is working as expected.

### Case 3
1. Go to the WooCommerce settings page: `wp-admin/admin.php?page=wc-settings`.
2. Change 'Default customer location' to `Shop country/region`. 
3. Add the store's address.
4. Add the rates that cover the store's address
5. As a guest/not logged-in user, visit the cart page with some products added already.
7. Confirm the shipping address calculator link to change or add a new address is displayed.

* [ ] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested (ie: it makes changes to tests, coding standards, docblocks, etc.). -->
* [ ] Should be tested by the development team exclusively <!-- Check this box if this PR should be tested by the development team exclusively (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

## Screenshots or screencast

<!-- Any screenshots of UI changes will be helpful to include here. Leave blank if not applicable. -->

| Before | After |
| ------ | ----- |
|![image](https://github.com/woocommerce/woocommerce-blocks/assets/11503784/dbe26987-038a-45ac-b71d-5ec1dc69d05d)|![image](https://github.com/woocommerce/woocommerce-blocks/assets/11503784/7c736bc1-594c-40f8-a527-c4b299c3765b)|

## WooCommerce Visibility

<!-- Check this documentation link (../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WooCommerce core, part of the feature plugin, or experimental. -->
Required:

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental
* [ ] N/A

## Checklist

Required:
* [x] This PR has either a `[type]` label or a `[skip-changelog]` label.
* [ ] This PR is assigned to a milestone.

Conditional:
* [ ] This PR has a changelog description (if `[skip-changelog]` label is not present).
* [ ] This PR adds/removes a feature flag & I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
* [ ] This PR adds/removes an experimental interfaces, and I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
* [ ] This PR has been accessibility tested.
* [ ] This PR has had any necessary documentation added/updated.

## Changelog
<!-- Provide a brief, descriptive summary of the changes in this PR. Include potential impacts on different parts of the product. Example: "Updated the checkout process to streamline the experience for users and reduce the number of steps." -->

> Display shipping calculator link for guests shoppers.

Commit:
Display shipping calculator when formatted address is present

- Earlier, shipping calculator was getting displayed based on the isAddressComplete value.
- For the scenarios where address was incomplete but formatted address was present shipping calculator was not getting displayed.
- This made shipping calculator not getting displayed for guest shoppers.
- With this, conditions are changed from isAddressComplete to formatShippingAddress to display shipping calculator if formatted address is present.
- Add unit test case for the condition for formatted address.

ralucaStan commented:
@tarunvijwani to be super safe add also testing steps where the shipping address has all fields empty and the Add and address link to see shipping options link appears. That would imply not defaulting shipping to the store address

ralucaStan commented on assets/js/base/components/cart-checkout/totals/shipping/shipping-address.tsx:
```diff
@@ -31,15 +28,17 @@ export const ShippingAddress = ( {
+	const hasFormattedAddress = !! formatShippingAddress( shippingAddress );
+
+	// If we don't have formatted address, and we're not in the editor, don't show anything.
+	if ( ! hasFormattedAddress && ! isEditor ) {
```
This comments is redundant as it explains the code, but does not provide context. What case are we actually covering with this and why are we returning nothing?

opr replied:
Yup, maybe the comment can explain _why_ `hasFormattedAddress` would be false? (i.e. the shopper hasn't entered one)

tarunvijwani replied:
@ralucaStan Thank you for the suggestion, I will update the comment and add more testing steps.

>What case are we actually covering with this and why are we returning nothing?

If there is no default customer location in the store and the customer has not entered their address, but there is a default shipping method available for all locations, then we will hide the shipping calculator. This is because the store already has a default shipping address. This change was made in response to this request: https://github.com/woocommerce/woocommerce-blocks/pull/8141#issuecomment-1397239622

Here are the steps to reproduce this behavior:

1. Go to the WooCommerce settings page: `wp-admin/admin.php?page=wc-settings`.
3. Change 'Default customer location' to `No location by default`. 
4. Add any default(fallback) shipping rates. 
5. Open your site in a private session(incognito mode).
6. Add some products to the cart. 
7. Go to the cart page, and confirm the shipping calculator is hidden and default shipping rates are visible
<img width="841" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/11503784/8daa2400-7436-4463-a08c-95b87ec92eb6">

<img width="1566" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/11503784/0110b41a-29ee-4b4e-97f9-b76abe2f6b59">
<img width="756" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/11503784/c9372b61-170b-4a65-9e10-f2528023b8d3">

senadir replied:
I understand that we have this rather complex sutation because we choose to have two buttons, one named "change address" and one named "enter an address to view shipping rates"? Otherwise I don't understand why we need to check if a formatted address is present to begin with to see the calculator.

tarunvijwani replied:
@senadir Yes, it wouldn't make sense to show the 'Add address for shipping rates' link when default rates are already displayed. Similarly, we should not display the 'Change address' link without an address. Hence we decided to hide it in this condition.

senadir replied:
What happens when you have rates but not an address? Do we show nothing at that step?

tarunvijwani replied:
@senadir No, we don't show a calculator in that scenario. It will be addressed in the https://github.com/woocommerce/woocommerce-blocks/issues/10327 ticket.

opr reviewed (approved):
Changes look fine besides the comment Raluca mentioned, a couple of the testing steps seem slightly redundant, why make the tester add rates for California, then update their store's address? We could just specify that there should be rates that cover the store's address couldn't we?

I tested with:
- above steps
- no shipping rates for the default location
- only local pickup for the default location

Since this works OK and the updates are only comments/testing steps I will approve.

Commit:
Update comments to explain the condition to hide shipping calculator

- Update the comments to add the reason to hide shipping calculator.
- When there is no default customer location in the store and the customer has not entered their address, but there is a default shipping method available for all locations, then we will hide the shipping calculator.

tarunvijwani commented:
@opr I have updated the testing steps and comments. I'd appreciate it if you could review them before I merge the PR. Thank you so much!

opr reviewed (changes requested):
Thanks I re-tested this with this configuration:

- no default location,
- location-specific methods _and_ fallback methods enabled

I feel that the calculator should be shown here. Why don't we show it in this case?

tarunvijwani commented:
> Thanks I re-tested this with this configuration:
> 
> * no default location,
> * location-specific methods _and_ fallback methods enabled
> 
> I feel that the calculator should be shown here. Why don't we show it in this case?

@opr There is a discussion ongoing about which text to show for the calculator link. Click here https://github.com/woocommerce/woocommerce-blocks/pull/11442#discussion_r1371500652 for more information.

I feel we should tackle that in a different PR. Happy to hear your thoughts about it.

opr commented:
@tarunvijwani - ok if you create a follow-up issue to tackle this then that should be OK!

opr reviewed (approved):
Approving on the basis that a follow-up ticket will be created to tackle the issue of the calculator not showing if:

- No location set by default,
- Location-specific methods _and_ fallback methods available

tarunvijwani commented:
> @opr I have updated the testing steps and comments. I'd appreciate it if you could review them before I merge the PR. Thank you so much!

We already have a separate ticket for it.  It will be addressed in the woocommerce/woocommerce#42280 ticket.

Commit:
Merge branch 'trunk' into fix/11435-display-shipping-calculator-for-guests
